bil.ott

```diff
@@ -506,13 +527,40 @@ defns reduce_exp :: '' ::=
  ----------------------------------------- :: bop_lhs
  delta |- e1 bop v2 ~> e1' bop v2
 
+ delta |- e2 ~> e2'
+ ------------------------------------------ :: sop_rhs
+ delta |- e1 sop e2 ~> e1 sop e2'
+
+ delta |- e1 ~> e1'
+ ----------------------------------------- :: sop_lhs
+ delta |- e1 sop v2 ~> e1' sop v2
+
+ delta |- e2 ~> e2'
+ ------------------------------------------ :: cop_rhs
+ delta |- e1 cop e2 ~> e1 cop e2'
+
+ delta |- e1 ~> e1'
+ ----------------------------------------- :: cop_lhs
+ delta |- e1 cop v2 ~> e1' cop v2
 
  -------------------------------------------------------- :: bop_unk_rhs
  delta |- e bop unknown[str]:t ~> unknown[str]:t
 
  -------------------------------------------------------- :: bop_unk_lhs
  delta |- unknown[str]:t bop e ~> unknown[str]:t
 
+ dom(delta) |- e :: t'
```

Comment:
We indeed do allow the right-hand side of the shift operator to have arbitrary width, this is however a totally unnecessary relaxation, that we should remove from the implementation. 

So, I'm proposing to require shift operands to have the same type (this will essentially remove the necessity of the syntactic class of shifts and bring shifts back to the bop class. 

Also, since we're factoring out comparison operators from binary operators, we can't still name the latter binary (as comparison operators are also binary). My suggestion would be to factor to split the class of binary operators into to sub-classes arithmetic operators `aop` and logical operators `lop`, e.g.,

```
bop ::= aop | lop
aop ::= + | - | .. | & | xor 
lop ::= < | = | ..
```

The bop class is not strictly necessary, but we can keep it for consistency with the  uop, and for clarity, I believe.

Reply:
This makes sense to me and is a bit cleaner than what I wrote up. I'll have a new commit momentarily.